rust/src/lib.rs

```diff
@@ -425,59 +450,58 @@ impl Toodle {
+            if item.due_date != due_date {
+                let due_date_kw = kw!(:item/due_date);
+                let due_date_ref = in_progress.get_entid(&due_date_kw).ok_or_else(|| ErrorKind::UnknownAttribute(due_date_kw))?;
+                if let Some(date) = due_date {
+                    let _ = builder.add(entid.clone(), due_date_ref, date.to_typed_value());
+                } else if let Some(date) = item.due_date {
+                    let _ = builder.retract(entid.clone(), due_date_ref, date.to_typed_value());
+                }
             }
-        }
 
-        if let Some(new_labels) = labels {
-            let existing_labels = self.fetch_labels_for_item(&(item.uuid)).unwrap_or(vec![]);
-            let labels_to_add = new_labels.iter()
-                                        .filter(|label| !existing_labels.contains(label) && label.id.is_some() )
-                                        .map(|label|  format!("{}", label.id.clone().map::<i64, _>(|e| e.into()).unwrap()) )
-                                        .collect::<Vec<String>>()
-                                        .join(", ");
-            if !labels_to_add.is_empty() {
-                transaction.push(format!("[:db/add {0} :item/label [{1}]]", &item_id.id, labels_to_add));
-            }
-            let labels_to_remove = existing_labels.iter()
-                                        .filter(|label| !new_labels.contains(label) && label.id.is_some() )
-                                        .map(|label|  format!("{}", label.id.clone().map::<i64, _>(|e| e.into()).unwrap()) )
-                                        .collect::<Vec<String>>()
-                                        .join(", ");
-            if !labels_to_remove.is_empty() {
-                transaction.push(format!("[:db/retract {0} :item/label [{1}]]", &item_id.id, labels_to_remove));
+            if item.completion_date != completion_date {
+                let completion_date_kw = kw!(:item/completion_date);
+                let completion_date_ref = in_progress.get_entid(&completion_date_kw).ok_or_else(|| ErrorKind::UnknownAttribute(completion_date_kw))?;
+                if let Some(date) = completion_date {
+                    let _ = builder.add(entid.clone(), completion_date_ref, date.to_typed_value());
+                } else if let Some(date) = item.completion_date {
+                    let _ = builder.retract(entid.clone(), completion_date_ref, date.to_typed_value());
+                }
             }
-        }
 
-        // TODO: better transact API.
-        let query = format!("[{0}]", transaction.join(""));
+            if let Some(new_labels) = labels {
+                let item_labels_kw = kw!(:item/label);
+                let item_labels_ref = in_progress.get_entid(&item_labels_kw).ok_or_else(|| ErrorKind::UnknownAttribute(item_labels_kw))?;
+                for label in new_labels {
+                    if !existing_labels.contains(&label) && label.id.is_some() {
```

Comment:
It's totally safe to re-transact a unique statement. Indeed, if you do

```
[-1 :label/name "thing"]
[-1 :label/color "red"]
```

then later 

```
[-2 :label/name "thing"]
[-2 :label/color "blue"]
```

we will transact only one thing:

```
[65591 :label/color "red" :retract]
[65591 :label/color "blue" :add]
```

You shouldn't need to check for existence here.